app/post_metadata.go

```diff
@@ -0,0 +1,418 @@
+
+	return names
+}
+
+func (a *App) getCustomEmojisForPost(post *model.Post, reactions []*model.Reaction) ([]*model.Emoji, *model.AppError) {
+	if !*a.Config().ServiceSettings.EnableCustomEmoji {
+		// Only custom emoji are returned
+		return []*model.Emoji{}, nil
+	}
+
+	names := getEmojiNamesForPost(post, reactions)
+
+	if len(names) == 0 {
+		return []*model.Emoji{}, nil
+	}
+
+	return a.GetMultipleEmojiByName(names)
+}
+
+// Given a string, returns the first autolinked URL in the string as well as an array of all Markdown
+// images of the form ![alt text](image url). Note that this does not return Markdown links of the
+// form [text](url).
+func getFirstLinkAndImages(str string) (string, []string) {
+	firstLink := ""
+	images := []string{}
+
+	markdown.Inspect(str, func(blockOrInline interface{}) bool {
+		switch v := blockOrInline.(type) {
+		case *markdown.Autolink:
+			if firstLink == "" {
+				firstLink = v.Destination()
+			}
+		case *markdown.InlineImage:
+			images = append(images, v.Destination())
+		case *markdown.ReferenceImage:
+			images = append(images, v.ReferenceDefinition.Destination())
+		}
+
+		return true
+	})
+
+	return firstLink, images
+}
+
+func getImagesInMessageAttachments(post *model.Post) []string {
+	var images []string
+
+	for _, attachment := range post.Attachments() {
+		_, imagesInText := getFirstLinkAndImages(attachment.Text)
+		images = append(images, imagesInText...)
+
+		_, imagesInPretext := getFirstLinkAndImages(attachment.Pretext)
+		images = append(images, imagesInPretext...)
+
+		for _, field := range attachment.Fields {
+			if value, ok := field.Value.(string); ok {
+				_, imagesInFieldValue := getFirstLinkAndImages(value)
+				images = append(images, imagesInFieldValue...)
+			}
+		}
+
+		if attachment.AuthorIcon != "" {
+			images = append(images, attachment.AuthorIcon)
+		}
+
+		if attachment.ImageURL != "" {
+			images = append(images, attachment.ImageURL)
+		}
+
+		if attachment.ThumbURL != "" {
+			images = append(images, attachment.ThumbURL)
+		}
+
+		if attachment.FooterIcon != "" {
+			images = append(images, attachment.FooterIcon)
+		}
+	}
+
+	return images
+}
+
+func (a *App) getLinkMetadata(requestURL string, useCache bool) (*opengraph.OpenGraph, *model.PostImage, error) {
+	// Check cache
+	if useCache {
+		og, image, ok := getLinkMetadataFromCache(requestURL)
+
+		if ok {
+			return og, image, nil
+		}
+	}
+
+	// Make request for a web page or an image
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	request.Header.Add("Accept", "text/html, image/*")
+
+	res, err := a.HTTPService.MakeClient(false).Do(request)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer consumeAndClose(res)
+
+	// Parse the data
+	og, image, err := a.parseLinkMetadata(requestURL, res.Body, res.Header.Get("Content-Type"))
+
+	// Write back to cache
+	if useCache {
+		cacheLinkMetadata(requestURL, og, image)
```

Comment:
When you get from the cache, I like returning both types since it allows us to cache the data together without making the calling code handle an `interface{}`, and I cached them together since it only requires checking a single cache when we see a link that we don't know what to do with.

I can see your point about the parameters for `cacheLinkMetadata` being a bit odd, but assuming we're still caching them together, I think it makes sense to do it this way. If we were to just pass in a single `interface{}`, we'd still need the logic somewhere to figure out whether or not to pass `og` or `image` into it. Or maybe we could use some common wrapper for the two although that would still likely require something that tells us what the actual type of the data is.